### ISSUE TITLE
queryString producing empty query on bound prepared statements

### ISSUE THREAD

**Issue by dmoston** [outworkers/phantom#647]
We have recently converted to prepared statements.
I am trying to log our generated queries with little success (hopefully I am not doing something stupid).
It seems that queryString always returns only the terminate semicolon ";".
I attempted to prove my point by adding the following failing test to the PreparedSelectQueryTest
```
  it should "generate queryString for bound prepared statements" in {
    val recipe = gen[Recipe]

    val query = database.recipes.select.where(_.url eqs ?).prepare()
    val bound = query.bind("URL")
    info(s"bound.queryString=${bound.queryString}")
    bound.queryString should include ("WHERE")
  }
```

**Comment by dmoston** [comment]
I am also seeing errors in our logs that appear to be attempting to log the failed BoundStatement, but it does not properly log complete queryString (only "BoundStatement@43b4e300").
Is it possible to log the complete text of the queryString?
```
2017-03-31 13:32:02.825 c.o.phantom [ERROR] Failed to execute query com.datastax.driver.core.BoundStatement@43b4e300
com.datastax.driver.core.exceptions.OperationTimedOutException: [***] Timed out waiting for server response
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onTimeout(RequestHandler.java:766) ~[stormjar.jar:?]
        at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1267) ~[stormjar.jar:?]
        at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581) ~[stormjar.jar:?]
        at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:655) ~[stormjar.jar:?]
        at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367) ~[stormjar.jar:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_72]
```
I am guessing that this comes from this here: https://github.com/outworkers/phantom/blob/cc07de125218951f6410971bdca6407d8a64f028/phantom-dsl/src/main/scala/com/outworkers/phantom/builder/query/CassandraOperations.scala#L75
```
      def onFailure(err: Throwable): Unit = {
        Manager.logger.error(s"Failed to execute query $st", err)
        promise failure err
      }
```

**Comment by alexflav23** [comment]
HI @dmoston,

We are working to address this, the problem is an internal inconsistency within the Java Driver, namely that the `.toString` call on a `BoundStatement` doesn't actually return the query, which is the case for every other statement.

The blocker has been having to completely replace the bind mechanism used internally by the Java driver as its very slow and inefficient when it comes to looking up types and it requires very hacky handling of custom types.

This is happening already on #646, but it will take a week for this to complete, at which point Phantom will be handling all aspects of preparing a statement and you will have a concise query log.

Regards.

**Comment by dmoston** [comment]
Looks like #646 https://github.com/outworkers/phantom/pull/646 has been closed.
Should we now be watching #648 https://github.com/outworkers/phantom/pull/648?
Do you have an ETA?

**Comment by alexflav23** [comment]
I don't have a better ETA than "whenever it's ready" as it's quite a nasty one, hard to know when complete. We are hoping for end of month at the very latest.

**Comment by alexflav23** [comment]
Hi @dmoston,

This will be fixed once #668 gets merged, that's a gigantic changeset but it replaces the mechanism using to bind values to queries which gives us control of how they get serialised to a string as well. The reason why the `queryString` method wasn't there is because when binding the values get directly serialised to a `ByteBuffer`, bypassing the entire conversion to a CQL query string.

I'm telling you this because calling the method to print out the query will not be "for free", you will pay the additional performance penalty of toString serialisation, but obviously I don't disagree this can be useful tool for debugging.

Regards.

**Comment by alexflav23** [comment]
Hi @dmoston,

We have added an optional debug flag for batch statements. This import has to be available in scope wherever you are calling `bind`, and it will cause the macro generating the parser bindings for prepared statements to produce a debug query.

```
import com.outworkers.phantom.macros.debug.options.ShowBoundStatements
```

As I previously stated in #687:

When you call bind, a macro is invoked, to automatically generate the necessary code to deserialize concrete types to bytes, e.gByteBuffer. If you want to produce a queryString, you need to drop some of that performance and call to string, and retrace that in the original query.

I'm making you aware you are paying twice the cost of serialisation, which may not be trivial in all cases. That aside, we can definitely implement an accurate queryString.

It's way those debug strings are not produced automatically, and instead you are allowed to control them.

Regards,
Flavian
